File: internal/engine/engine.go
package engine

import (
	"beautifulmind/internal/deck"
	"beautifulmind/internal/embedding"
	"beautifulmind/internal/storage"
	"math"
	"math/rand"
	"sort"
	"time"
)

type KnowledgeGraph struct {
	Concepts map[string]*storage.DBConcept
	Cards    map[string]*storage.DBCard
	Links    []deck.Link
	Decks    []*deck.Deck
}

type JourneyStep struct {
	Card              *storage.DBCard
	AnsweredCorrectly bool
}

type SynapseSession struct {
	DB               *storage.DB
	Graph            *KnowledgeGraph
	Thesis           *storage.DBCard
	EvidenceDeck     []*storage.DBCard
	CurrentStep      int
	JourneyLog       []JourneyStep
	Score            int
	IsEvidenceRevealed bool
}

func LoadKnowledgeGraph(db *storage.DB, decks []*deck.Deck) (*KnowledgeGraph, error) {
	concepts, err := db.GetAllConceptsWithEmbeddings()
	if err != nil { return nil, err }

	allCards, err := db.GetAllCards()
	if err != nil { return nil, err }

	graph := &KnowledgeGraph{
		Concepts: make(map[string]*storage.DBConcept),
		Cards:    make(map[string]*storage.DBCard),
		Decks:    decks,
	}

	for _, c := range concepts { graph.Concepts[c.ID] = c }
	for _, c := range allCards { graph.Cards[c.ID] = c }
	for _, d := range decks { graph.Links = append(graph.Links, d.Links...) }

	return graph, nil
}

func (kg *KnowledgeGraph) NewSynapseSession(db *storage.DB, thesisCardID string) *SynapseSession {
	thesis := kg.Cards[thesisCardID]
	var evidenceDeck []*storage.DBCard
	for _, link := range kg.Links {
		if link.Target == thesis.ID {
			evidenceDeck = append(evidenceDeck, kg.Cards[link.Source])
		}
	}
	
	rand.Seed(time.Now().UnixNano())
	rand.Shuffle(len(evidenceDeck), func(i, j int) { evidenceDeck[i], evidenceDeck[j] = evidenceDeck[j], evidenceDeck[i] })

	return &SynapseSession{
		DB:           db,
		Graph:        kg,
		Thesis:       thesis,
		EvidenceDeck: evidenceDeck,
	}
}

func (s *SynapseSession) ProcessInput(key string) {
	card := s.CurrentCard()
	if card == nil { return }
	
	isReasoningCard := s.CurrentCardIsReasoning()

	if !s.IsEvidenceRevealed {
		if key == "enter" {
			s.IsEvidenceRevealed = true
			// if it's just a remembering card, we auto-grade it here
			if !isReasoningCard {
				s.updateSRS(card, 3) // grade 'good' just for revealing
			}
		}
	} else {
		if isReasoningCard {
			// handle reasoning answer
			choice := deck.LinkType("")
			if key == "up" { choice = deck.SUPPORTS }
			if key == "down" { choice = deck.REFUTES }
			
			if choice != "" {
				correctChoice := s.getCorrectLinkType()
				isCorrect := choice == correctChoice
				
				var quality int
				if isCorrect { s.Score += 100; quality = 4 } else { quality = 1 } // easy or hard
				
				s.JourneyLog = append(s.JourneyLog, JourneyStep{Card: card, AnsweredCorrectly: isCorrect})
				s.updateSRS(card, quality)
				s.advance()
			}
		} else {
			// handle remembering card advance
			if key == "enter" {
				// already graded on reveal, so just log and advance
				s.JourneyLog = append(s.JourneyLog, JourneyStep{Card: card, AnsweredCorrectly: true})
				s.advance()
			}
		}
	}
}

// helper to advance to the next card
func (s *SynapseSession) advance() {
	s.CurrentStep++
	s.IsEvidenceRevealed = false
}

func (s *SynapseSession) CurrentCard() *storage.DBCard {
	if s.CurrentStep >= len(s.EvidenceDeck) { return nil }
	return s.EvidenceDeck[s.CurrentStep]
}

// the magic function: dynamically checks what kind of card we're on
func (s *SynapseSession) CurrentCardIsReasoning() bool {
	return s.getCorrectLinkType() != ""
}

func (s *SynapseSession) getCorrectLinkType() deck.LinkType {
	card := s.CurrentCard()
	if card == nil { return "" }
	for _, link := range s.Graph.Links {
		if link.Source == card.ID && link.Target == s.Thesis.ID {
			// only supports and refutes are reasoning tasks in this context
			if link.Type == deck.SUPPORTS || link.Type == deck.REFUTES {
				return link.Type
			}
		}
	}
	return ""
}

func (s *SynapseSession) updateSRS(card *storage.DBCard, quality int) {
	dbCard, err := s.DB.GetCard(card.ID)
	if err != nil { return }

	if quality < 3 {
		dbCard.Repetitions = 0; dbCard.Interval = 1
	} else {
		dbCard.Repetitions++
		if dbCard.Repetitions == 1 { dbCard.Interval = 1 } else if dbCard.Repetitions == 2 { dbCard.Interval = 6 } else {
			dbCard.Interval = math.Ceil(dbCard.Interval * dbCard.EaseFactor)
		}
	}
	if quality >= 3 {
		dbCard.EaseFactor += (0.1 - (5-float64(quality))*(0.08+(5-float64(quality))*0.02))
		if dbCard.EaseFactor < 1.3 { dbCard.EaseFactor = 1.3 }
	}

	duration := time.Duration(dbCard.Interval*24) * time.Hour
	dbCard.DueDate = time.Now().Add(duration)

	s.DB.UpdateCardSRS(dbCard.ID, dbCard.DueDate, dbCard.Interval, dbCard.EaseFactor, dbCard.Repetitions)
}

func (s *SynapseSession) GetRelatedConcepts() []string {
	card := s.CurrentCard()
	if card == nil { return nil }

	currentConcept := s.Graph.Concepts[card.ConceptID]
	type conceptSimilarity struct{ Name string; Sim float32 }
	var similarities []conceptSimilarity
	
	for _, otherConcept := range s.Graph.Concepts {
		if otherConcept.ID != currentConcept.ID {
			sim := embedding.CosineSimilarity(currentConcept.Embedding, otherConcept.Embedding)
			if sim > 0.1 {
				similarities = append(similarities, conceptSimilarity{Name: otherConcept.Name, Sim: sim})
			}
		}
	}
	sort.Slice(similarities, func(i, j int) bool { return similarities[i].Sim > similarities[j].Sim })

	var relatedNames []string
	for i, sim := range similarities {
		if i >= 4 { break }
		relatedNames = append(relatedNames, sim.Name)
	}
	return relatedNames
}

File: node/src/engine.ts
import { DB, Card, Concept, Correlation } from './db';

type Link = { source: string; target: string; type: string };

export class SynapseSession {
    private db: DB;
    private links: Link[];
    public thesis: Card;
    public evidenceDeck: Card[];
    public currentStep: number;
    public isEvidenceRevealed: boolean;
    public isInfinite: boolean;
    private visitedCardIds: Set<string>;
    public sprintTotal: number;
    public sprintCompleted: number;

    constructor(db: DB, decksDir: string, thesisId: string, isInfinite: boolean) {
        this.db = db;
        this.links = this.db.getLinks(decksDir);
        this.thesis = this.db.getCard(thesisId)!;
        this.isInfinite = isInfinite;
        this.currentStep = 0;
        this.isEvidenceRevealed = false;
        
        this.evidenceDeck = this.links
            .filter(link => link.target === this.thesis.id)
            .map(link => this.db.getCard(link.source))
            .filter((card): card is Card => card !== null && new Date(card.due_date) <= new Date());
        
        this.visitedCardIds = new Set(this.evidenceDeck.map(c => c.id));
        this.visitedCardIds.add(this.thesis.id);
        
        this.sprintTotal = this.evidenceDeck.length;
        this.sprintCompleted = 0;

        for (let i = this.evidenceDeck.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [this.evidenceDeck[i], this.evidenceDeck[j]] = [this.evidenceDeck[j], this.evidenceDeck[i]];
        }
    }

    public processInput(key: string) {
        const card = this.getCurrentCard();
        if (!card) return;

        const isReasoningCard = this.links.some(l => l.source === card.id && l.target === this.thesis.id);

        if (!this.isEvidenceRevealed) {
            if (key === 'enter') {
                this.isEvidenceRevealed = true;
                if (!isReasoningCard) this.updateSRS(card, 3);
            }
        } else {
            let advance = false;
            if (isReasoningCard) {
                let choice: string | null = null;
                if (key === 'up') choice = 'SUPPORTS';
                if (key === 'down') choice = 'REFUTES';
                
                if (choice) {
                    const link = this.links.find(l => l.source === card.id && l.target === this.thesis.id);
                    const isCorrect = choice === link?.type;
                    this.updateSRS(card, isCorrect ? 4 : 1);
                    advance = true;
                }
            } else {
                if (key === 'enter') {
                    advance = true;
                }
            }
            if (advance) {
                if(this.currentStep < this.sprintTotal) this.sprintCompleted++;
                this.advance();
            }
        }
    }
    
    private advance() {
        this.currentStep++;
        this.isEvidenceRevealed = false;
        if (this.currentStep >= this.evidenceDeck.length && this.isInfinite) {
            this.extendJourney();
        }
    }

    private extendJourney() {
        const lastCard = this.evidenceDeck[this.evidenceDeck.length - 1];
        if (!lastCard) return;

        const correlations = this.db.getCorrelationsForConcept(lastCard.concept_id);
        
        for (const corr of correlations) {
            const nextCard = this.db.getNextCardForConcept(corr.id, Array.from(this.visitedCardIds));
            if (nextCard) {
                this.evidenceDeck.push(nextCard);
                this.visitedCardIds.add(nextCard.id);
                return;
            }
        }
    }

    public getCurrentCard(): Card | null {
        return this.evidenceDeck[this.currentStep] || null;
    }
    
    private updateSRS(card: Card, quality: number) {
        let { repetitions, interval, ease_factor } = card;

        if (quality < 3) {
            repetitions = 0;
            interval = 1;
        } else {
            repetitions++;
            if (repetitions === 1) interval = 1;
            else if (repetitions === 2) interval = 6;
            else interval = Math.ceil(interval * ease_factor);
        }

        ease_factor = ease_factor + (0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02));
        if (ease_factor < 1.3) ease_factor = 1.3;

        const dueDate = new Date();
        dueDate.setDate(dueDate.getDate() + interval);
        
        this.db.updateCardSRS(card.id, {
            dueDate: dueDate.toISOString(),
            interval,
            easeFactor: ease_factor,
            repetitions,
        });
    }

    public getState() {
        return {
            thesis: this.thesis,
            evidenceDeck: this.evidenceDeck,
            currentStep: this.currentStep,
            isEvidenceRevealed: this.isEvidenceRevealed,
            currentCard: this.getCurrentCard(),
            sprintTotal: this.sprintTotal,
            sprintCompleted: this.sprintCompleted
        };
    }
}